[WFLY-10522] Support Eclipse MicroProfile Config

jmesnil opened the pull request:
Add the microprofile-config-smallrye subsystem that provides support for
the Eclipse MicroProfile Config using the SmallRye implementation at
https://github.com/smallrye/smallrye-config.

The integrated version is 1.3.0.Final that supports MicroProfile Config
1.3.

The name of the subsystem is `microprofile-config-smallrye`.
It is added by default to *all* WildFly configuration (standalone and
domain).

JIRA: https://issues.jboss.org/browse/WFLY-10522

---

Please make sure your PR meets the following requirements:
- [x] Pull Request title is properly formatted: `[WFLY-XYZ] Subject` or `WFLY-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue(s)
- [x] Pull Request contains description of the issue(s)
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Attached commits represent units of work and are properly formatted

For bigger changes, major and minor component upgrades make sure your PR also meets following requirements:
- [x] Pull Request requires a change to the documentation
- [x] Documentation have been updated accordingly
=> the PR contains the subsystem documentation in the admin guide
- [x] Tests were added to cover changes
=> integration tests checks that deployments can use the MicroProfile Config API
- [x] Analysis was done in wildfly/wildfly-proposals#79
- [x] Test Plan has been done
- [x] Tests were verified in advance

jmesnil committed:
[WFLY-10522] Support Eclipse MicroProfile Config

Add the microprofile-config-smallrye subsystem that provides support for
the Eclipse MicroProfile Config using the SmallRye implementation at
https://github.com/smallrye/smallrye-config.

The integrated version is 1.3.3 that supports MicroProfile Config
1.3.

The name of the subsystem is `microprofile-config-smallrye`.
It is added by default to standalone WildFly configurations.

JIRA: https://issues.jboss.org/browse/WFLY-10522

jmesnil commented:
- [x] Analysis was done in https://github.com/wildfly/wildfly-proposals/pull/79

jmesnil commented:
retest this please

jmesnil commented:
The failures are related to the PR.
* I'm working on fixing the mixed-domain test
* the smallrye-config implementation seems to have issue when the security manager is enabled (it's not using privileged action to get the TCCL).

jmesnil commented:
retest this please

marekkopecky commented:
Is it necessary to reload server after one new property is added?

```
[standalone@localhost:9990 /] /subsystem=microprofile-config-smallrye/config-source=aa:add()
{"outcome" => "success"}
[standalone@localhost:9990 /] /subsystem=microprofile-config-smallrye/config-source=aa:write-attribute(name=properties,value={"a"="b","c"="d"})
{
    "outcome" => "success",
    "response-headers" => {
        "operation-requires-reload" => true,
        "process-state" => "reload-required"
    }
}
[standalone@localhost:9990 /] reload 
[standalone@localhost:9990 /] /subsystem=microprofile-config-smallrye/config-source=aa:map-put(name=properties,key=e,value=e)
{
    "outcome" => "success",
    "response-headers" => {
        "operation-requires-reload" => true,
        "process-state" => "reload-required"
    }
}
[standalone@localhost:9990 /] 
```

marekkopecky commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc:
```diff
@@ -0,0 +1,106 @@
+</subsystem>
+----
+
+=== ConfigSource from Directory
```
It seems that it is also possible to define ConfigSource by "class" name, right? Can you please describe it in this documentation?
```
/subsystem=microprofile-config-smallrye/config-source=aaaa:add(class={
```

jmesnil replied:
done

marekkopecky commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc:
```diff
@@ -0,0 +1,106 @@
+This corresponds to the layout used by https://docs.openshift.com/enterprise/3.2/dev_guide/configmaps.html[OpenShift ConfigMaps].
+The `dir` value corresponds to the `mountPath` in the ConfigMap definition in OpenShift or Kubernetes.
+
+
```
Can you please describe also `/subsystem=microprofile-config-smallrye/config-source-provider=` settings in documentation?

jmesnil replied:
done

jmesnil commented:
@marekkopecky yes,the server has to be reloaded when properties are modified in the config-source resources for the time being.

The smallrye-config is copying the properties when an application is deployed. Updating the properties in the resource will not update the ConfigSource used by the deployments.

There are enhancements planned for next MicroProfile Config specification to be more dynamic so that ConfigSource can notify their content has changed. When this will be implemented in smallrye-config, reloading the server will no longer be required.

wolfc commented:
Why is this not coming through a new feature pack provisioned by Galleon?

jmesnil commented:
This new extension depends on the weld (and ee) extensions that are provided by WildFly main feature-pack.

In order to be provided by a separate feature pack, it would first be required to provide separate feature pack for Weld (some kind of baseline for microprofile spec). This goes beyond the scope of this RFE.

jmesnil committed:
[WFLY-10522] Support Eclipse MicroProfile Config

* in domain.xml, exclude the extension from previous WildFly releases
* remove the subsystem and extension in DomainAdjuster710 for
  mixed-domain tests

JIRA: https://issues.jboss.org/browse/WFLY-10522

jmesnil commented:
I have fixed the failures that were related to the PR. I don't expect anymore change before it can be reviewed.

jstourac commented on testsuite/integration/basic/src/test/java/org/wildfly/test/integration/microprofile/config/smallrye/AssertUtils.java:
```diff
@@ -0,0 +1,36 @@
+public class AssertUtils {
+
+    public static void assertTextContainsProperty(String text, String propName, Object propValue) {
+        Assert.assertTrue("Property " + propName + " not found in " + text,
```
Maybe wrap `text` variable into some quotes so it's more easily distinguishable where compared text starts?
```
"Property " + propName + " not found in '" + text + "'"
```

jstourac commented:
@jmesnil would you mind rebasing this agains master? We would like to add MR for you with some other tests. It would be nice to avoid any possible discrepancies that could arrive in testsuite changes in the meantime. Thank you.

jmesnil committed:
[WFLY-10522] Support Eclipse MicroProfile Config

* Improve resource description

JIRA: https://issues.jboss.org/browse/WFLY-10522

jmesnil committed:
[WFLY-10522] Support Eclipse MicroProfile Config

* Attach the ConfigProviderService to the deployment to be able to
  release its config when the server is shutdown.

JIRA: https://issues.jboss.org/browse/WFLY-10522
JIRA: https://issues.jboss.org/browse/WFWIP-64

bstansberry commented on microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/ConfigSourceDefinition.java:
```diff
@@ -0,0 +1,154 @@
+            .setAttributeMarshaller(AttributeMarshaller.ATTRIBUTE_OBJECT)
+            .setRestartAllServices()
+            .build();
+    static AttributeDefinition DIR = SimpleAttributeDefinitionBuilder.create("dir", ModelType.STRING)
```
Filesystem path attributes should be paired with a 'relative-to'.

jmesnil replied:
@bstansberry I have fixed it in a dedicated commit (and check that file system completion works)

rsvoboda commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc:
```diff
@@ -0,0 +1,155 @@
+This extension is included in the standard configurations included in the
+WildFly distribution.
+
+You can also add the extension to a configuration without it either by adding
```
So MP config won't be enabled by default ?

jmesnil replied:
I mention just above that 
> This extension is included in the standard configurations included in the WildFly distribution

Is the wording ambiguous?

rsvoboda commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc:
```diff
@@ -0,0 +1,155 @@
+[NOTE]
+This corresponds to the layout used by https://docs.openshift.com/enterprise/3.2/dev_guide/configmaps.html[OpenShift ConfigMaps].
+The `dir` value corresponds to the `mountPath` in the ConfigMap definition in OpenShift or Kubernetes.
+
```
Please add note that deployment must have CDI enabled - e.g. by META-INF/beans.xml or CDI bean defining annotation.

jmesnil replied:
I have updated the doc

jmesnil commented:
@jstourac I've rebased

mjurc commented:
@jmesnil The last rebase made `wildfly` unbuildable:
`[ERROR] Failed to execute goal on project wildfly-galleon-pack: Could not resolve dependencies for project org.wildfly:wildfly-galleon-pack:pom:14.0.0.Beta2-SNAPSHOT: Failure to find org.wildfly:wildfly-microprofile-config-smallrye:jar:14.0.0.Beta2-SNAPSHOT in http://repository.jboss.org/nexus/content/groups/public/ was cached in the local repository, resolution will not be reattempted until the update interval of jboss-public-repository-group has elapsed or updates are forced -> [Help 1]
`

mjurc commented on microprofile/config-smallrye/pom.xml:
```diff
@@ -0,0 +1,93 @@
+        Maintain separation between the artifact id and the version to help prevent
+        merge conflicts between commits changing the GA and those changing the V.
+        -->
+        <version>14.0.0.Beta1-SNAPSHOT</version>
```
This needs to be bumped to `14.0.0.Beta2-SNAPSHOT`, else `wildfly` is unbuildable

jmesnil commented on microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/ConfigProviderService.java:
```diff
@@ -0,0 +1,57 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2017, Red Hat, Inc., and individual contributors
```
Most of the code was taken from https://github.com/wildfly-extras/wildfly-microprofile-config/tree/1.x/extension. I preserved their file headers (including the copyright date)

jmesnil committed:
[WFLY-10522] Support Eclipse MicroProfile Config

Fix dir attribute of config-source so that it uses WildFly
path features (including relative-to and completion from CLI)

JIRA: https://issues.jboss.org/browse/WFLY-10522
JIRA: https://issues.jboss.org/browse/WFWIP-58

tremes commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc:
```diff
@@ -0,0 +1,157 @@
+[NOTE]
+This corresponds to the layout used by https://docs.openshift.com/enterprise/3.2/dev_guide/configmaps.html[OpenShift ConfigMaps].
+The `dir` value corresponds to the `mountPath` in the ConfigMap definition in OpenShift or Kubernetes.
+
```
This seems to me a little bit poor description regarding OpenShift integration. So does this mean that only properties configured in ConfigMap will be supported?

Env. variables defined e.g in Pod or DeploymentConfig will be ignored?

jmesnil commented:
retest this please

bstansberry commented on microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/DependencyProcessor.java:
```diff
@@ -0,0 +1,78 @@
+        final ModuleLoader moduleLoader = Module.getBootModuleLoader();
+
+        moduleSpecification.addSystemDependency(new ModuleDependency(moduleLoader, MICROPROFILE_CONFIG_API, false, false, true, false));
+        moduleSpecification.addSystemDependency(new ModuleDependency(moduleLoader, MICROPROFILE_CONFIG_EXTENSION, false, false, true, false));
```
Exposing a subsystem module to a deployment isn't good. We should get clean separation between deployment code and management code, e.g. so we can do things like https://docs.google.com/document/d/1OK8izSS9DVSOAqXjadG7piIjWOJ4KU2eKQ548ZL26aM/edit#heading=h.n28b42dwjj5x

jmesnil replied:
fixed. the deployments no longer has any dependency on the extension module.

bstansberry commented on feature-pack/src/main/resources/configuration/domain/template.xml:
```diff
@@ -96,20 +97,31 @@
             </excluded-extensions>
         </host-exclude>
+        <!--
+        <host-exclude name="WildFly13.0">
```
https://github.com/wildfly/wildfly-core/pull/3416

jmesnil replied:
I will uncomment that in a subsequent PR as there is not yet a Core release with that key

jmesnil commented:
retest this please
retest this please

mjurc committed:
[WFLY-10522] Test extension to check various data types and simple priority checks

* various data types are check that they work as expected in regards to default values, overridden values and empty values (no default)
* simple smoke checks for priority when properties are loaded from different sources
* checks for Converter and ConfigSource implementations loaded by service loader and their priority.

jmesnil commented:
retest this please

jmesnil commented:
The PR has been updated to integrate tests written by @mjurc (thanks!)

@wildfly/mergers-full I've removed the missing-reqs flag. 

It is ready to be reviewed & merged as it as has addressed all reported issues or requested changes.

jamezp commented on microprofile/config-smallrye/pom.xml:
```diff
@@ -0,0 +1,93 @@
+        </dependency>
+        <dependency>
+            <groupId>org.jboss.logging</groupId>
+            <artifactId>jboss-logging-annotations</artifactId>
```
This should also include:
```
<!-- This is a compile-time dependency of this project, but is not needed at compile or runtime by other
      projects that depend on this project.-->
<scope>provided</scope>
<optional>true</optional>
```

jmesnil replied:
fixed

jamezp commented on microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/ConfigSourceDefinition.java:
```diff
@@ -0,0 +1,168 @@
+            .setAllowNull(true)
+            .setAttributeMarshaller(AttributeMarshaller.ATTRIBUTE_OBJECT)
+            .setRestartAllServices()
+            .setCapabilityReference("org.wildfly.management.path-manager")
```
Should this reference be here or on the `DIR` attribute?

jmesnil replied:
fixed

jamezp commented on microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/DependencyProcessor.java:
```diff
@@ -0,0 +1,78 @@
+ */
+public class DependencyProcessor implements DeploymentUnitProcessor {
+
+    Logger log = Logger.getLogger(SubsystemDeploymentProcessor.class);
```
Is this used anywhere? I'm not seeing it used unless I'm missing something. Having it be named `SubsystemDeploymentProcessor.class` seems a bit odd too.

jmesnil replied:
removed

jamezp commented on microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/SubsystemDeploymentProcessor.java:
```diff
@@ -0,0 +1,117 @@
+     * The current number is large enough for it to happen after all
+     * the standard deployment unit processors that come with JBoss AS.
+     */
+    public static final int PRIORITY = 0x4000;
```
Are we no longer putting these in the `Phase`?

jmesnil replied:
@jamezp Good catch, I set it like that when the code was in wildfly-extras repo and was not "officially" in WildFly code base.

Is it acceptable to use existing priority + offset (e.g. `Phase.POST_MODULE_WELD_PORTABLE_EXTENSIONS + 10`) ?

jamezp commented on microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/DependencyProcessor.java:
```diff
@@ -0,0 +1,78 @@
+     * The current number is large enough for it to happen after all
+     * the standard deployment unit processors that come with JBoss AS.
+     */
+    public static final int PRIORITY = 0x4000;
```
Are we no longer putting these in the `Phase`?

jmesnil replied:
this one should simply be Phase.DEPENDENCIES_MODULE

jamezp reviewed (changes requested):
It's probably not a big deal, but there are some test classes that are missing the copyright headers.

jmesnil committed:
[WFLY-10522] Support Eclipse MicroProfile Config

* Fix deployment phases
* add missing header to test files

JIRA: https://issues.jboss.org/browse/WFLY-10522

jmesnil commented:
@jamezp I also added the missing copyright headers.

The last thing that may required a change is the priority of the processor. Are you ok with using existing priority + offset (`Phase.POST_MODULE_WELD_PORTABLE_EXTENSIONS + 10`) or must I add an explicit priority to the Phase in wildfly-core (in that case, this PR must be hold until there is a core release with that new Phase).

jmesnil commented:
retest this please

jmesnil commented:
@jamezp I've created https://issues.jboss.org/browse/WFCORE-4017 to make sure I don't forget to update the DUP's priority

jamezp commented:
Perfect thanks @jmesnil. We can make these changes later in the cycle since it's more of a fix than a feature.

jamezp reviewed (approved):
I'm approving this and we can deal with the Phase stuff later.
